lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -2195,9 +2212,16 @@ def contourf(self, X, Y, Z, *args, zdir='z', offset=None, **kwargs):
 
         jX, jY, jZ = art3d.rotate_axes(X, Y, Z, zdir)
         cset = super().contourf(jX, jY, jZ, *args, **kwargs)
-        self.add_contourf_set(cset, zdir, offset)
-
-        self.auto_scale_xyz(X, Y, Z, had_data)
+        levels = self.add_contourf_set(cset, zdir, offset)
+
+        # Autoscale in the zdir based on the levels added, which are
+        # different from data range if any contour extensions are present
+        args = {'x': X, 'y': Y, 'z': Z, zdir: levels}
+        # Input data and levels have different sizes, but auto_scale_xyz
+        # expected same-size input, so manually take min/max limits
+        args = {key: [np.nanmin(vals), np.nanmax(vals)]
+                for key, vals in args.items()}
```

Comment:
Change naming to make this more telling:
```suggestion
        # Autoscale in the zdir based on the levels added, which are
        # different from data range if any contour extensions are present
        dimension_values = {'x': X, 'y': Y, 'z': Z, zdir: levels}
        # Input data and levels have different sizes, but auto_scale_xyz
        # expected same-size input, so manually take min/max limits
        limits = {dimension: [np.nanmin(vals), np.nanmax(vals)]
                for dimension, vals in dim_values.items()}
```

Reply:
Also done. I also re-did this a bit to remove the redundant dict (use a list instead), and moved it to a helper method so it can be used with `tricontourf` too.